File: lesunny/054_[1차]다트게임.java
class Solution {
    public int solution(String dartResult) {
        int answer = 0;
        int[] games = new int[3];
        int idx = 0;
        
        for(int i=0; i<3 ; i++){
            while(Character.isDigit(dartResult.charAt(idx))){
                System.out.print(dartResult.charAt(idx)+" ");
                games[i]*=10;
                games[i]+=Integer.parseInt(dartResult.substring(idx,idx+1));

                idx++;
            }
            
            try{
                if(dartResult.substring(idx,idx+1).equals("S")){
                    games[i] = (int)Math.pow(games[i],1);
                }else if(dartResult.substring(idx,idx+1).equals("D")){
                    games[i] = (int)Math.pow(games[i],2);
                }else if(dartResult.substring(idx,idx+1).equals("T")){
                    games[i] = (int)Math.pow(games[i],3);
                }
                idx++;
            }catch(StringIndexOutOfBoundsException e){
            }
            
            try{
                if(dartResult.substring(idx,idx+1).equals("#")){
                    games[i] *= -1;
                }else if(dartResult.substring(idx,idx+1).equals("*")){
                    games[i] *= 2;
                    if(i!=0)
                        games[i-1]*=2;
                }else if(idx < dartResult.length() && Character.isDigit(dartResult.charAt(idx))){
                    continue;
                }else{
                    continue;

                }
                idx++;
            }catch(StringIndexOutOfBoundsException e){
            }
        }

        for(int i=0; i<3; i++){
            answer += games[i];
        }
        
        return answer;
    }
}

File: JIWON/054 [1차] 다트 게임.cpp
#include <string>
#include <vector>

using namespace std;

int solution(string dartResult) {
    int answer = 0;
    vector<int> v;
    int s = 0;
    
    for(int i=0;i<dartResult.size();i++){
        if(dartResult[i] >= '0' && dartResult[i] <= '9'){
            s = dartResult[i] - '0';
            if(dartResult[i] == '1' && dartResult[i+1] == '0'){
                s = 10;
                i++;
            }
        }
        
        if(dartResult[i] == 'S') v.push_back(s), s = 0;
        else if(dartResult[i] == 'D') v.push_back(s*s), s = 0;
        else if(dartResult[i] == 'T') v.push_back(s*s*s), s = 0;
        else if(dartResult[i] == '*'){
            if(v.size() > 1) {
                v.back() *= 2;
                v[v.size() - 2] *= 2;
            }
            else v.back() *= 2;
        }
        else if(dartResult[i] == '#'){
            v.back() = -v.back();
        }
    }
    
    for(int i=0;i<v.size();i++){
        answer += v[i];
    }
    
    return answer;
}
